src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs

```diff
@@ -0,0 +1,129 @@
+﻿// ------------------------------------------------------------------------------
+// 
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+// 
+// This code is licensed under the MIT License.
+// 
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+// 
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace Microsoft.Identity.Client.ApiConfig
+{
+    // TODO: we should only have values in here publicly that are useful for ALL acquire token usages
+    // (e.g. WithExtraQueryParameters).  Which ones don't make the cut and need to be moved up the 
+    // inheritance hierarchy?
+    /// <summary>
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <typeparam name="TParams"></typeparam>
+    public abstract class AbstractAcquireTokenParameterBuilder<T, TParams>
+        where T : AbstractAcquireTokenParameterBuilder<T, TParams>
+        where TParams : IAcquireTokenCommonParameters
+    {
+        internal AcquireTokenParameters Parameters { get; } = new AcquireTokenParameters();
+
+        /// <summary>
+        /// </summary>
+        /// <param name="scopes"></param>
+        /// <returns></returns>
+        protected T WithScopes(IEnumerable<string> scopes)
+        {
+            Parameters.Scopes = scopes;
+            return (T)this;
+        }
+
```

Comment:
Would we want to have .WithResource(string resource), which in effect could do a scope of resource/.default?
probably not ?

Reply:
discussed with jm and others, not doing this.

---
In reply to: [246267901](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246267901) [](ancestors = 246267901)